fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -2954,6 +2956,121 @@ public void validateMetaData() {
+     *         a value other than {@link SplitHelper#UNSPLIT_RECORD}, the {@link FDBStoreTimer.Counts#INVALID_SPLIT_SUFFIX}
+     *         counter will be incremented.
+     *     <li>For record stores in which record splitting is disabled, but the record key is longer than expected,
+     *         the {@link FDBStoreTimer.Counts#INVALID_KEY_LENGTH} counter will be incremented.
+     * </ul>
+     *
+     * @param continuation continuation from a previous repair attempt or {@code null} to start from the beginning
+     * @param scanProperties properties to provide scan limits on the repair process
+     *   record keys should be logged
+     * @return a continuation or {@code null} if the repair has been completed
+     */
+    public CompletableFuture<byte[]> repairRecordKeys(@Nullable byte[] continuation, @Nonnull ScanProperties scanProperties) {
+        // If the records aren't split to begin with, then there is nothing to do.
+        if (getRecordMetaData().isSplitLongRecords()) {
+            return CompletableFuture.completedFuture(null);
+        }
+
+        if (scanProperties.getExecuteProperties().getIsolationLevel().isSnapshot()) {
+            throw new RecordCoreArgumentException("Cannot repair record key split markers at SNAPSHOT isolation level")
+                    .addLogInfo(LogMessageKeys.SCAN_PROPERTIES, scanProperties);
+        }
+
+        final Subspace recordSubspace = recordsSubspace();
+        KeyValueCursor cursor = KeyValueCursor.Builder.withSubspace(recordSubspace)
+                .setContext(getRecordContext())
+                .setContinuation(continuation)
+                .setScanProperties(scanProperties)
+                .build();
+
+        final AtomicReference<byte[]> nextContinuation = new AtomicReference<>();
+        final FDBRecordContext context = getContext();
+        return AsyncUtil.whileTrue( () ->
+                cursor.onNext().thenApply( result -> {
+                    if (!result.hasNext()) {
+                        if (result.hasStoppedBeforeEnd()) {
+                            nextContinuation.set(result.getContinuation().toBytes());
+                        }
+                        return false;
+                    }
+
+                    repairRecordKeyIfNecessary(context, recordSubspace, result.get());
+                    return true;
+                })).thenApply(ignored -> nextContinuation.get());
+    }
+
+    private void repairRecordKeyIfNecessary(@Nonnull FDBRecordContext context, @Nonnull Subspace recordSubspace, @Nonnull KeyValue keyValue) {
+        final RecordMetaData metaData = metaDataProvider.getRecordMetaData();
+        final Tuple recordKey = recordSubspace.unpack(keyValue.getKey());
+
+        // Ignore version key
+        if (metaData.isStoreRecordVersions() && isMaybeVersion(recordKey)) {
+            return;
+        }
+
+        final Message record = serializer.deserialize(metaData, recordKey, keyValue.getValue(), getTimer());
```

Comment:
Will any serializers be confused by being given the whole key as the primary key?

Reply:
That's a good question, but I'm not sure what I can do here; I could go with the approach that, if the last element of the tuple is a `0L`, then lop it off and treat that as the PK. However, until I know the type of record it is I cannot be sure as to whether that `0L` was actually part fo the primary key, but the split suffix was missing or if it was a legit split suffix.  I'm open to suggestions on which approach might be safest.

Reply:
Let's stick with this and see if it becomes an issue.